README.md
# Sonata Doctrine Phpcr Admin Bundle

Symfony Sonata / Integrate Doctrine PHPCR into the SonataAdminBundle

[![Latest Stable Version](https://poser.pugx.org/sonata-project/doctrine-phpcr-admin-bundle/v/stable)](https://packagist.org/packages/sonata-project/doctrine-phpcr-admin-bundle)
[![Latest Unstable Version](https://poser.pugx.org/sonata-project/doctrine-phpcr-admin-bundle/v/unstable)](https://packagist.org/packages/sonata-project/doctrine-phpcr-admin-bundle)
[![License](https://poser.pugx.org/sonata-project/doctrine-phpcr-admin-bundle/license)](https://packagist.org/packages/sonata-project/doctrine-phpcr-admin-bundle)

[![Total Downloads](https://poser.pugx.org/sonata-project/doctrine-phpcr-admin-bundle/downloads)](https://packagist.org/packages/sonata-project/doctrine-phpcr-admin-bundle)
[![Monthly Downloads](https://poser.pugx.org/sonata-project/doctrine-phpcr-admin-bundle/d/monthly)](https://packagist.org/packages/sonata-project/doctrine-phpcr-admin-bundle)
[![Daily Downloads](https://poser.pugx.org/sonata-project/doctrine-phpcr-admin-bundle/d/daily)](https://packagist.org/packages/sonata-project/doctrine-phpcr-admin-bundle)

Branch | Travis | Coveralls | Scrutinizer |
------ | ------ | --------- | ----------- |
2.x   | [![Build Status][travis_stable_badge]][travis_stable_link]     | [![Coverage Status][coveralls_stable_badge]][coveralls_stable_link]     | [![Scrutinizer Status][scrutinizer_stable_badge]][scrutinizer_stable_link] |
master | [![Build Status][travis_unstable_badge]][travis_unstable_link] | [![Coverage Status][coveralls_unstable_badge]][coveralls_unstable_link] | [![Scrutinizer Status][scrutinizer_unstable_badge]][scrutinizer_unstable_link] |

## Documentation

Check out the documentation on the [official website](https://sonata-project.org/bundles/doctrine-phpcr-admin).

## Support

For general support and questions, please use [StackOverflow](http://stackoverflow.com/questions/tagged/sonata).

If you think you found a bug or you have a feature idea to propose, feel free to open an issue
**after looking** at the [contributing guide](CONTRIBUTING.md).

## License

This package is available under the [MIT license](LICENSE).

[travis_stable_badge]: https://travis-ci.org/sonata-project/SonataDoctrinePhpcrAdminBundle.svg?branch=2.x
[travis_stable_link]: https://travis-ci.org/sonata-project/SonataDoctrinePhpcrAdminBundle
[travis_unstable_badge]: https://travis-ci.org/sonata-project/SonataDoctrinePhpcrAdminBundle.svg?branch=master
[travis_unstable_link]: https://travis-ci.org/sonata-project/SonataDoctrinePhpcrAdminBundle

[coveralls_stable_badge]: https://coveralls.io/repos/github/sonata-project/SonataDoctrinePhpcrAdminBundle/badge.svg?branch=2.x
[coveralls_stable_link]: https://coveralls.io/github/sonata-project/SonataDoctrinePhpcrAdminBundle?branch=2.x
[coveralls_unstable_badge]: https://coveralls.io/repos/github/sonata-project/SonataDoctrinePhpcrAdminBundle/badge.svg?branch=master
[coveralls_unstable_link]: https://coveralls.io/github/sonata-project/SonataDoctrinePhpcrAdminBundle?branch=master

[scrutinizer_stable_badge]: https://scrutinizer-ci.com/g/sonata-project/SonataDoctrinePhpcrAdminBundle/badges/quality-score.png?b=2.x
[scrutinizer_stable_link]: https://scrutinizer-ci.com/g/sonata-project/SonataDoctrinePhpcrAdminBundle/?branch=2.x
[scrutinizer_unstable_badge]: https://scrutinizer-ci.com/g/sonata-project/SonataDoctrinePhpcrAdminBundle/badges/quality-score.png?b=master
[scrutinizer_unstable_link]: https://scrutinizer-ci.com/g/sonata-project/SonataDoctrinePhpcrAdminBundle/?branch=master


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## [2.3.0](https://github.com/sonata-project/SonataDoctrinePhpcrAdminBundle/compare/2.2.0...2.3.0) - 2019-10-03
### Fixed
- `Child "_delete" does not exist` in Collection when we don't have delete permission

### Changed
- `sonata-project/core-bundle` dependencies from `^3.8` to `^3.14`
- use new `Sonata\Form\Type\*` namespace of SonataCoreBundle

## [2.2.0](https://github.com/sonata-project/SonataDoctrinePhpcrAdminBundle/compare/2.1.1...2.2.0) - 2019-01-22

### Added
- Symfony 4 support

### Fixed
- Form Type Filter ChoiceType Symfony 3 compatibility
- Deprecation for symfony/config 4.2+

### Removed
- support for php 5 and php 7.0

## [2.1.1](https://github.com/sonata-project/SonataDoctrinePhpcrAdminBundle/compare/2.1.0...2.1.1) - 2018-02-08
### Changed
- Switch all templates references to Twig namespaced syntax
- Switch from templating service to sonata.templating

## [2.1.0](https://github.com/sonata-project/SonataDoctrinePhpcrAdminBundle/compare/2.0.0...2.1.0) - 2018-01-08
### Added
- version 2.0 for `phpcr-odm`

### Removed
- Support for old versions of PHP and Symfony.

### Fixed
- Missing root_node parameter in cmfTree options

###### Starting this point, the changelog does not follow the same way.

2.0.0
-----

* **2017-04-26**: [BC break] The tree routing is no longer optional. Require `@SonataDoctrinePhpcrAdminBundle/Resources/config/routing/tree.xml` in your routing file.
* **2017-02-06**: [BC break] Removed `Datagrid\SimplePager` in favor of the one provided by the SonataAdminBundle.
* **2015-05-06**: [BC break] Removed `Tree\PhpcrOdmTree`.
* **2015-05-06**: [BC break] Changed first argument of `Controller\TreeController` from `TreeInterface $tree` to `$repositoryName = 'default'`.
* **2015-05-06**: [BC break] The tree block template now only recieves a `repository_name`, `root_node` and `routing_defaults` variables.
* **2015-05-06**: Added `document_tree_repository` setting.
* **2015-05-06**: [BC break] Removed `Form\Type\TreeModelType#setTree()` method and `$tree` property.

1.2.0
-----

* **2014-10-22**: Updated the template `Block/tree.html.twig`, parameters are now wrapped in `settings`.
* **2014-10-13**: Fixed things around showing relations
* **2014-09-29**: Now compatible with sonata admin 2.3

1.2.0-RC1
---------

* **2014-08-20**: Updated to PSR-4 autoloading

1.1.0
---------

Release 1.1.0

2014-04-11
----------

- drop Symfony 2.2 compatibility

1.1.0-RC1
---------

2014-03-24
----------

- Updated the dependencies on sonata to use the new SonataCoreBundle.

2014-02-14
----------

- The tree options confirm_move and depth are now grouped in the configuration
  `document_tree_options`, plus there is an additional option
  `precise_children` to further help with flat trees.

1.0.1
-----

2013-11-12
----------

- Added support for the Sonata ACL editor, if that is enabled in the configuration.
  http://www.sonata-project.org/bundles/admin/master/doc/reference/security.html#acl-editor

1.0.0
-----

2013-09-24
----------

- Improved the document tree configuration so that "all" works as it was intended.
  Added validation of the class names used in the tree - if you start seeing
  exceptions that means you have invalid class names in your `document_tree`.

2013-07-27
----------

- Changed show URL from `/<doc-path>` to `<doc-path>/show` as this agrees with the Sonata path.

2013-02-14
----------

Renamed form type `doctrine_phpcr_type_tree_model` to `doctrine_phpcr_odm_tree`. You will need to update your form types / admin classes, e.g. change:

    $builder->add('parent', 'doctrine_phpcr_type_tree_model');
    // to
    $builder->add('parent', 'doctrine_phpcr_odm_tree');

2013-08-06
----------

renamed:    Resources/views/CRUD/edit_orm_many_association_script.html.twig     ->   Resources/views/CRUD/edit_phpcr_many_association_script.html.twig
renamed:    Resources/views/CRUD/edit_orm_many_to_many.html.twig                ->   Resources/views/CRUD/edit_phpcr_many_to_many.html.twig
renamed:    Resources/views/CRUD/edit_orm_many_to_one.html.twig                 ->   Resources/views/CRUD/edit_phpcr_many_to_one.html.twig
renamed:    Resources/views/CRUD/edit_orm_one_association_script.html.twig      ->   Resources/views/CRUD/edit_phpcr_one_association_script.html.twig
renamed:    Resources/views/CRUD/edit_orm_one_to_many.html.twig                 ->   Resources/views/CRUD/edit_phpcr_one_to_many.html.twig
renamed:    Resources/views/CRUD/edit_orm_one_to_one.html.twig                  ->   Resources/views/CRUD/edit_phpcr_one_to_one.html.twig
renamed:    Resources/views/CRUD/list_orm_many_to_many.html.twig                ->   Resources/views/CRUD/list_phpcr_many_to_many.html.twig
renamed:    Resources/views/CRUD/list_orm_many_to_one.html.twig                 ->   Resources/views/CRUD/list_phpcr_many_to_one.html.twig
renamed:    Resources/views/CRUD/list_orm_one_to_many.html.twig                 ->   Resources/views/CRUD/list_phpcr_one_to_many.html.twig
renamed:    Resources/views/CRUD/list_orm_one_to_one.html.twig                  ->   Resources/views/CRUD/list_phpcr_one_to_one.html.twig

    You should change any template extensions accordingly: {% extends 'SonataDoctrinePHPCRAdminBundle:CRUD:...' %}

Renamed blocks in Resources/views/Form/form_admin_fields.html.twig You should change any block overrides accordingly.


Tree/PhpcrOdmTree.php
<?php

/*
 * This file is part of the Sonata Project package.
 *
 * (c) Thomas Rabaix <thomas.rabaix@sonata-project.org>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

namespace Sonata\DoctrinePHPCRAdminBundle\Tree;

use Doctrine\Common\Util\ClassUtils;
use Doctrine\ODM\PHPCR\Document\Generic;
use Doctrine\ODM\PHPCR\DocumentManager;
use PHPCR\Util\NodeHelper;
use PHPCR\Util\PathHelper;
use Sonata\AdminBundle\Admin\AdminInterface;
use Sonata\AdminBundle\Admin\Pool;
use Sonata\DoctrinePHPCRAdminBundle\Model\ModelManager;
use Symfony\Bundle\FrameworkBundle\Templating\Helper\AssetsHelper;
use Symfony\Cmf\Bundle\TreeBrowserBundle\Tree\TreeInterface;
use Symfony\Component\PropertyAccess\Exception\NoSuchPropertyException;
use Symfony\Component\PropertyAccess\PropertyAccess;
use Symfony\Component\Templating\Helper\CoreAssetsHelper;
use Symfony\Component\Translation\TranslatorInterface;

/**
 * A tree implementation to work with Doctrine PHPCR-ODM.
 *
 * Your documents need to map all children with an Children mapping for the
 * tree to see its children. Not having the Children annotation is a
 * possibility to not show children you do not want to show.
 *
 * @author David Buchmann <david@liip.ch>
 * @author Uwe Jäger <uwej711@googlemail.com>
 */
class PhpcrOdmTree implements TreeInterface
{
    /**
     * @var ModelManager
     */
    private $defaultModelManager;

    /**
     * @var DocumentManager
     */
    private $dm;

    /**
     * @var Pool
     */
    private $pool;

    /**
     * @var TranslatorInterface
     */
    private $translator;

    /**
     * @var CoreAssetsHelper
     */
    private $assetHelper;

    /**
     * Array of cached admin services indexed by class name.
     *
     * @var array
     */
    private $admins = array();

    /**
     * List of the valid class names that may be used as tree "ref" fields.
     *
     * @var array
     */
    private $validClasses;

    /**
     * Depth to which grand children should be fetched, currently the maximum depth is one.
     *
     * @var int
     */
    private $depth;

    /**
     * Fetch children lazy - enabling this will allow the tree to fetch a larger amount of  children in the tree but less accurate.
     *
     * @var bool
     */
    private $preciseChildren;

    /**
     * The options are.
     *
     * - depth: Down to what level children should be fetched, currently the
     *      maximum supported depth is one.
     * - precise_children: To determine if a tree element has children, check if
     *      the document has valid children. If false, simply check if the node
     *      has any child nodes. Less accurate but better performance.
     *
     * @param DocumentManager               $dm
     * @param ModelManager                  $defaultModelManager to use with documents that have no manager
     * @param Pool                          $pool                to get admin classes for documents from
     * @param TranslatorInterface           $translator
     * @param CoreAssetsHelper|AssetsHelper $assetHelper
     * @param array                         $validClasses        list of the valid class names that may be
     *                                                           used as tree "ref" fields
     *                                                           $param integer $depth depth to which grand children should be fetched,
     *                                                           currently the maximum depth is one
     * @param array                         $options
     */
    public function __construct(
        DocumentManager $dm,
        ModelManager $defaultModelManager,
        Pool $pool,
        TranslatorInterface $translator,
        $assetHelper,
        array $validClasses,
        array $options
    ) {
        $this->dm = $dm;
        $this->defaultModelManager = $defaultModelManager;
        $this->pool = $pool;
        $this->translator = $translator;
        $this->assetHelper = $assetHelper;
        $this->validClasses = $validClasses;

        $this->depth = $options['depth'];
        $this->preciseChildren = $options['precise_children'];

        if (!$this->assetHelper instanceof CoreAssetsHelper) {
            // check for the AssetsHelper class introduced in Symfony 2.7
            if (!$this->assetHelper instanceof AssetsHelper) {
                throw new \InvalidArgumentException(sprintf(
                    'The asset helper input for the %s class is not valid.'
                    .' The current asset helper is an instance of %s.'
                    .' It should be an instance of one of the following classes: %s',
                    __CLASS__,
                    get_class($this->assetHelper),
                    implode(', ', array(
                        'Symfony\Component\Templating\Helper\CoreAssetsHelper',
                        'Symfony\Bundle\FrameworkBundle\Templating\Helper\AssetsHelper',
                    ))
                ));
            }
        }
    }

    /**
     * Get the children of the document at this path by looking at the Child and Children mappings.
     *
     * {@inheritdoc}
     */
    public function getChildren($path)
    {
        $root = $this->dm->find(null, $path);

        $children = array();

        if ($root) {
            $rootManager = $this->getModelManager($root);
            foreach ($this->getDocumentChildren($rootManager, $root) as $document) {
                if ($document instanceof Generic &&
                    (NodeHelper::isSystemItem($document->getNode())
                        || !strncmp('phpcr_locale:', $document->getNode()->getName(), 13)
                    )
                ) {
                    continue;
                }
                $manager = $this->getModelManager($document);

                $child = $this->documentToArray($manager, $document);

                if ($this->depth > 0) {
                    foreach ($this->getDocumentChildren($manager, $document) as $grandchild) {
                        $child['children'][] = $this->documentToArray($manager, $grandchild);
                    }
                }

                $children[] = $child;
            }
        }

        return $children;
    }

    /**
     * {@inheritdoc}
     */
    public function move($movedPath, $targetPath)
    {
        $resultingPath = $targetPath.'/'.basename($movedPath);

        $document = $this->dm->find(null, $movedPath);
        if (null === $document) {
            return "No document found at $movedPath";
        }

        $this->dm->move($document, $resultingPath);
        $this->dm->flush();

        $admin = $this->getAdmin($document);
        if (null !== $admin) {
            $id = $admin->getNormalizedIdentifier($document);
            $urlSafeId = $admin->getUrlsafeIdentifier($document);
        } else {
            $id = $this->defaultModelManager->getNormalizedIdentifier($document);
            $urlSafeId = $this->defaultModelManager->getUrlsafeIdentifier($document);
        }

        return array('id' => $id, 'url_safe_id' => $urlSafeId);
    }

    /**
     * @param object $document
     * @param object $child
     *
     * @return bool TRUE if valid, FALSE if not valid
     */
    public function isValidDocumentChild($document, $child)
    {
        $className = ClassUtils::getClass($document);
        $childClassName = ClassUtils::getClass($child);

        if (!isset($this->validClasses[$className])) {
            // no mapping means no valid children
            return false;
        }

        return in_array($childClassName, $this->validClasses[$className]['valid_children']);
    }

    /**
     * {@inheritdoc}
     */
    public function reorder($parent, $moved, $target, $before)
    {
        $parentDocument = $this->dm->find(null, $parent);
        $this->dm->reorder($parentDocument, basename($moved), basename($target), $before);
        $this->dm->flush();
    }

    /**
     * {@inheritdoc}
     */
    public function getAlias()
    {
        return 'phpcr_odm_tree';
    }

    /**
     * {@inheritdoc}
     */
    public function getNodeTypes()
    {
        $result = array();

        foreach ($this->validClasses as $className => $children) {
            $rel = $this->normalizeClassname($className);
            $admin = $this->getAdminByClass($className);
            $validChildren = array();

            foreach ($children['valid_children'] as $child) {
                $validChildren[] = $this->normalizeClassname($child);
            }

            $icon = 'bundles/cmftreebrowser/images/folder.png';
            if (!empty($children['image'])) {
                $icon = $children['image'];
            }

            $routes = array();
            if (null !== $admin) {
                foreach ($admin->getRoutes()->getElements() as $code => $route) {
                    $action = explode('.', $code);
                    $key = $this->mapAction(end($action));

                    if (null !== $key) {
                        $routes[$key] = sprintf('%s_%s', $admin->getBaseRouteName(), end($action));
                    }
                }
            }

            $result[$rel] = array(
                'icon' => array('image' => $this->assetHelper->getUrl($icon)),
                'label' => (null !== $admin) ? $admin->trans($admin->getLabel()) : $className,
                'valid_children' => $validChildren,
                'routes' => $routes,
            );
        }

        return $result;
    }

    /**
     * {@inheritdoc}
     */
    public function getLabels()
    {
        return array(
            'createItem' => $this->translator->trans('create_item', array(), 'SonataDoctrinePHPCRAdmin'),
            'deleteItem' => $this->translator->trans('delete_item', array(), 'SonataDoctrinePHPCRAdmin'),
        );
    }

    /**
     * Returns an array representation of the document.
     *
     * @param ModelManager $manager  the manager to use with this document
     * @param object       $document
     *
     * @return array
     */
    protected function documentToArray(ModelManager $manager, $document)
    {
        $className = ClassUtils::getClass($document);

        $rel = (in_array($className, array_keys($this->validClasses))) ? $className : 'undefined';
        $rel = $this->normalizeClassname($rel);

        $admin = $this->getAdmin($document);
        if (null !== $admin) {
            $label = $admin->toString($document);
            $id = $admin->getNormalizedIdentifier($document);
            $urlSafeId = $admin->getUrlsafeIdentifier($document);
        } else {
            $label = method_exists($document, '__toString') ? (string) $document : ClassUtils::getClass($document);
            $id = $manager->getNormalizedIdentifier($document);
            $urlSafeId = $manager->getUrlsafeIdentifier($document);
        }

        if (substr($label, 0, 1) === '/') {
            $label = PathHelper::getNodeName($label);
        }

        // TODO: ideally the tree should simply not make the node clickable
        $label .= $admin ? '' : ' '.$this->translator->trans('not_editable', array(), 'SonataDoctrinePHPCRAdmin');

        $hasChildren = false;
        if (isset($this->validClasses[$className]['valid_children'])
            && count($this->validClasses[$className]['valid_children'])
        ) {
            if ($this->preciseChildren) {
                // determine if a node has children the accurate way. we need to
                // loop over all documents, as a PHPCR node might have children but
                // only invalid ones. this is quite costly.
                $hasChildren = (bool) count($this->getDocumentChildren($manager, $document));
            } else {
                // just check if there is any child node
                $hasChildren = $manager->getDocumentManager()->getNodeForDocument($document)->hasNodes();
            }
        }

        return array(
            'data' => $label,
            'attr' => array(
                'id' => $id,
                'url_safe_id' => $urlSafeId,
                'rel' => $rel,
            ),
            'state' => $hasChildren ? 'closed' : null,
        );
    }

    /**
     * @param object $document
     * @param array  $children
     *
     * @return array of valid children for the document
     */
    protected function filterDocumentChildren($document, array $children)
    {
        $me = $this;

        return array_filter($children, function ($child) use ($me, $document) {
            return $me->isValidDocumentChild($document, $child);
        });
    }

    /**
     * @param object $document
     *
     * @return ModelManager the modelmanager for $document or the default manager
     */
    protected function getModelManager($document = null)
    {
        $admin = $document ? $this->getAdmin($document) : null;

        return $admin ? $admin->getModelManager() : $this->defaultModelManager;
    }

    /**
     * @param object $document the PHPCR-ODM document to get the sonata admin for
     *
     * @return AdminInterface
     */
    private function getAdmin($document)
    {
        $className = ClassUtils::getClass($document);

        return $this->getAdminByClass($className);
    }

    /**
     * @param string $className
     *
     * @return AdminInterface
     */
    private function getAdminByClass($className)
    {
        if (!isset($this->admins[$className])) {
            // will return null if not defined
            $this->admins[$className] = $this->pool->getAdminByClass($className);
        }

        return $this->admins[$className];
    }

    /**
     * @param ModelManager $manager  the manager to use with this document
     * @param object       $document the PHPCR-ODM document to get the children of
     *
     * @return array of children indexed by child nodename pointing to the child documents
     */
    private function getDocumentChildren(ModelManager $manager, $document)
    {
        $accessor = PropertyAccess::getPropertyAccessor(); // use deprecated BC method to support symfony 2.2

        /** @var $meta \Doctrine\ODM\PHPCR\Mapping\ClassMetadata */
        $meta = $manager->getMetadata(ClassUtils::getClass($document));

        $children = array();
        foreach ($meta->childrenMappings as $fieldName) {
            try {
                $prop = $accessor->getValue($document, $fieldName);
            } catch (NoSuchPropertyException $e) {
                $prop = $meta->getReflectionProperty($fieldName)->getValue($document);
            }
            if (null === $prop) {
                continue;
            }
            if (!is_array($prop)) {
                $prop = $prop->toArray();
            }

            $children = array_merge($children, $this->filterDocumentChildren($document, $prop));
        }

        foreach ($meta->childMappings as $fieldName) {
            try {
                $prop = $accessor->getValue($document, $fieldName);
            } catch (NoSuchPropertyException $e) {
                $prop = $meta->getReflectionProperty($fieldName)->getValue($document);
            }
            if (null !== $prop && $this->isValidDocumentChild($document, $prop)) {
                $children[$fieldName] = $prop;
            }
        }

        return $children;
    }

    /**
     * @param string $className
     *
     * @return string
     */
    private function normalizeClassname($className)
    {
        return str_replace('\\', '_', $className);
    }

    /**
     * @param string $action
     *
     * @return string|null
     */
    private function mapAction($action)
    {
        switch ($action) {
            case 'edit': return 'select_route';
            case 'create': return 'create_route';
            case 'delete': return 'delete_route';
        }

        return;
    }
}
